File: scripts/python/src/tools/core/native.py
"""Resolve the native executable built for this checkout."""

import json
import os
import shutil
from pathlib import Path

from tools.core.paths import repo_root


def binary(name):
    manifest = os.environ.get("DOTFILE_DEV_BUILD_MANIFEST")
    if manifest:
        selected = None
        for line in Path(manifest).read_text().splitlines():
            artifact = json.loads(line)
            if (
                artifact.get("reason") == "compiler-artifact"
                and artifact.get("target", {}).get("name") == name
                and artifact.get("executable")
            ):
                selected = Path(artifact["executable"])
        if selected and selected.is_file() and os.access(selected, os.X_OK):
            return str(selected)
        raise RuntimeError(f"prepared native binary missing: {name}")
    root = Path(repo_root())
    candidates = [
        root / "scripts/rust/target/debug" / name,
        root / "scripts/rust/target/release" / name,
        Path.home() / ".local/bin" / name,
    ]
    available = [path for path in candidates if path.is_file() and os.access(path, os.X_OK)]
    if available:
        return str(max(available, key=lambda path: path.stat().st_mtime))
    if found := shutil.which(name):
        return found
    raise RuntimeError(f"{name} is not built; run ./setup.sh")

File: scripts/rust/crates/workstation/tests/native.rs
#![forbid(unsafe_code)]

use std::fs;
use std::path::Path;
use workstation::native::Resolver;

fn native(path: &Path) {
    fs::create_dir_all(path.parent().unwrap()).unwrap();
    fs::write(path, b"\x7fELFfixture").unwrap();
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(path, fs::Permissions::from_mode(0o755)).unwrap();
    }
}

#[test]
fn executable_sibling_wins_and_fallbacks_ignore_modification_times() {
    let temp = tempfile::tempdir().unwrap();
    let mut resolver = Resolver {
        root: temp.path().join("repo"),
        home: temp.path().join("home"),
        current_exe: None,
        manifest: None,
    };
    let debug = resolver.root.join("scripts/rust/target/debug/tool");
    let release = resolver.root.join("scripts/rust/target/release/tool");
    let installed = resolver.home.join("dotfiles/.bin/tool");
    for path in [&release, &installed, &debug] {
        native(path);
    }
    assert_eq!(resolver.resolve("tool").unwrap(), Some(release.clone()));
    resolver.current_exe = Some(resolver.home.join("dotfiles/.bin/dotfile"));
    assert_eq!(resolver.resolve("tool").unwrap(), Some(installed.clone()));
    resolver.current_exe = Some(
        resolver
            .root
            .join("scripts/rust/target/debug/deps/test-123"),
    );
    assert_eq!(resolver.resolve("tool").unwrap(), Some(debug.clone()));
    fs::remove_file(debug).unwrap();
    assert_eq!(resolver.resolve("tool").unwrap(), Some(release.clone()));
    fs::remove_file(release).unwrap();
    assert_eq!(resolver.resolve("tool").unwrap(), Some(installed));
}

#[test]
fn prepared_manifest_is_authoritative_and_test_harnesses_are_not_tools() {
    let temp = tempfile::tempdir().unwrap();
    let manifest = temp.path().join("build.jsonl");
    let resolver = Resolver {
        root: temp.path().join("repo"),
        home: temp.path().join("home"),
        current_exe: None,
        manifest: Some(manifest.clone()),
    };
    native(&resolver.home.join(".local/bin/count"));
    let prepared = temp.path().join("prepared command");
    native(&prepared);
    let harness = temp.path().join("test harness");
    native(&harness);
    fs::write(&manifest, format!("{}\n{}\n", serde_json::json!({"reason":"compiler-artifact","target":{"name":"git-discard"},"profile":{"test":false},"executable":prepared}), serde_json::json!({"reason":"compiler-artifact","target":{"name":"git-discard"},"profile":{"test":true},"executable":harness}))).unwrap();
    assert_eq!(resolver.resolve("gdd").unwrap(), Some(prepared));
    assert_eq!(resolver.resolve("count").unwrap(), None);
    assert!(resolver.resolve("../count").is_err());
}

#[test]
fn interpreter_scripts_and_nonexecutables_are_excluded() {
    let temp = tempfile::tempdir().unwrap();
    let resolver = Resolver {
        root: temp.path().join("repo"),
        home: temp.path().join("home"),
        current_exe: None,
        manifest: None,
    };
    let launcher = resolver.home.join(".local/bin/sysinfo");
    native(&launcher);
    fs::write(
        &launcher,
        "#!/usr/bin/env python3\nraise RuntimeError('never launch')\n",
    )
    .unwrap();
    assert!(resolver.resolve("sysinfo").unwrap().is_none());
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        native(&launcher);
        fs::set_permissions(&launcher, fs::Permissions::from_mode(0o644)).unwrap();
        assert!(resolver.resolve("sysinfo").unwrap().is_none());
    }
}
